libsolidity/formal/CVC4Interface.cpp

```diff
@@ -120,66 +124,80 @@ CVC4::Expr CVC4Interface::toCVC4Expr(Expression const& _expr)
-	else if (n == "<")
-		return m_context.mkExpr(CVC4::kind::LT, arguments[0], arguments[1]);
-	else if (n == "<=")
-		return m_context.mkExpr(CVC4::kind::LEQ, arguments[0], arguments[1]);
-	else if (n == ">")
-		return m_context.mkExpr(CVC4::kind::GT, arguments[0], arguments[1]);
-	else if (n == ">=")
-		return m_context.mkExpr(CVC4::kind::GEQ, arguments[0], arguments[1]);
-	else if (n == "+")
-		return m_context.mkExpr(CVC4::kind::PLUS, arguments[0], arguments[1]);
-	else if (n == "-")
-		return m_context.mkExpr(CVC4::kind::MINUS, arguments[0], arguments[1]);
-	else if (n == "*")
-		return m_context.mkExpr(CVC4::kind::MULT, arguments[0], arguments[1]);
-	else if (n == "/")
-		return m_context.mkExpr(CVC4::kind::INTS_DIVISION_TOTAL, arguments[0], arguments[1]);
-	else if (n == "mod")
-		return m_context.mkExpr(CVC4::kind::INTS_MODULUS, arguments[0], arguments[1]);
-	else if (n == "select")
-		return m_context.mkExpr(CVC4::kind::SELECT, arguments[0], arguments[1]);
-	else if (n == "store")
-		return m_context.mkExpr(CVC4::kind::STORE, arguments[0], arguments[1], arguments[2]);
-	// Cannot reach here.
-	solAssert(false, "");
-	return arguments[0];
+		solAssert(_expr.hasCorrectArity(), "");
+		if (n == "ite")
+			return arguments[0].iteExpr(arguments[1], arguments[2]);
+		else if (n == "not")
+			return arguments[0].notExpr();
+		else if (n == "and")
+			return arguments[0].andExpr(arguments[1]);
+		else if (n == "or")
+			return arguments[0].orExpr(arguments[1]);
+		else if (n == "implies")
+			return m_context.mkExpr(CVC4::kind::IMPLIES, arguments[0], arguments[1]);
+		else if (n == "=")
+			return m_context.mkExpr(CVC4::kind::EQUAL, arguments[0], arguments[1]);
+		else if (n == "<")
+			return m_context.mkExpr(CVC4::kind::LT, arguments[0], arguments[1]);
+		else if (n == "<=")
+			return m_context.mkExpr(CVC4::kind::LEQ, arguments[0], arguments[1]);
+		else if (n == ">")
+			return m_context.mkExpr(CVC4::kind::GT, arguments[0], arguments[1]);
+		else if (n == ">=")
+			return m_context.mkExpr(CVC4::kind::GEQ, arguments[0], arguments[1]);
+		else if (n == "+")
+			return m_context.mkExpr(CVC4::kind::PLUS, arguments[0], arguments[1]);
+		else if (n == "-")
+			return m_context.mkExpr(CVC4::kind::MINUS, arguments[0], arguments[1]);
+		else if (n == "*")
+			return m_context.mkExpr(CVC4::kind::MULT, arguments[0], arguments[1]);
+		else if (n == "/")
+			return m_context.mkExpr(CVC4::kind::INTS_DIVISION_TOTAL, arguments[0], arguments[1]);
+		else if (n == "mod")
+			return m_context.mkExpr(CVC4::kind::INTS_MODULUS, arguments[0], arguments[1]);
+		else if (n == "select")
+			return m_context.mkExpr(CVC4::kind::SELECT, arguments[0], arguments[1]);
+		else if (n == "store")
+			return m_context.mkExpr(CVC4::kind::STORE, arguments[0], arguments[1], arguments[2]);
+
+		solAssert(false, "");
+	}
+	catch (CVC4::TypeCheckingException const& _e)
+	{
+		solAssert(false, _e.what());
+	}
+	catch (CVC4::Exception const& _e)
```

Comment:
Same here - catching ``CVC4::Exception`` only should be enough, shouldn't it? Here we need to be careful as not to catch ``std::exception``, though, since we'll otherwise catch the ``solAssert(false, "")`` in the end of the ``try`` block... maybe that should be moved *outside* the try block anyways?

Reply:
you mean `solAssert` should be moved outside?

Reply:
Yeah, I think the ``solAssert`` in the end of the try block should be moved outside the try block at the very end of the function (after the ``catch``es). If "things are fine", then you'll return inside the try block, if the try block throws, you end up in a catch and assert there, if you just didn't have any match, you move beyond the try block to the final assert.

That way we don't need to be concerned about which exceptions to catch and accidentally catching the ``solAssert`` in the try block itself (which would e.g. happen, if you caught ``std::exception``).

Reply:
Using the same argument from https://github.com/ethereum/solidity/pull/6894/files#r290697800, do you still prefer it outside or inside this catch?

Reply:
I'll add one more assert at the very end of the function.

Reply:
Well, another assert at the end of the function isn't really needed if you keep the one in L191... but since you have other solAsserts earlier in the try block and even nested try blocks anyways, I guess this is fine...

I'm not entirely convinced that this is the nicest way to do things, but it's not too important :-).